cpp/src/arrow/dataset/scanner.h

```diff
@@ -165,6 +166,9 @@ class ARROW_DS_EXPORT Scanner {
   /// Scan result in memory before creating the Table.
   Result<std::shared_ptr<Table>> ToTable();
 
+  /// \brief ToBatches returns an iterator over all Batches yielded by this scan.
+  Result<RecordBatchIterator> ToBatches();
```

Comment:
Nit: If I see `ToBatches` I expect `RecordBatchVector`.  I had named it `ScanBatches` but I don't feel too strongly on this point.

Reply:
You have a point. I didn't want to take ScanBatches since your method has a different signature. But maybe that could be `ScanIndexedBatches` or something.

Reply:
Hmm, the issue I ran into was that `ScanBatches` was used by `FileSystemDataset::Write` and it needed the fragment info in order to have access to the fragment's partition expression.  So at a minimum I needed to return "record batch & partition it came from".

I think there was some discussion (either on the ML or some JIRA/PR) about the benefit of keeping the fragment available as the user might want to know where the batch came from.

Can you modify the `ScanBatches` here to return a RecordBatch/Fragment pair?  I could align my `ScanBatches` with that (`PositionedRecordBatch` is overkill).